File: web_frontend/src/voice/jarvisVoice.ts
/**
 * JARVIS'S VOICE -- frontend side.
 *
 * One module, one voice. Before this, the chat bubble's speaker button
 * built its own utterance (rate 1.05, pitch 0.95), the call screen made
 * another, and the device TTS used a third set of numbers. Three voices
 * for one character, and any tuning meant hunting all three down.
 *
 * Everything that speaks now goes through speak() here, and the profile
 * comes from the server (/api/voice/profile), which reads the same
 * config/voice.json that the Python side uses. Tune it once, it changes
 * everywhere -- which is what "baad mein fine tune ho sake" requires.
 *
 * The amplitude callback is what drives the waveform on the voice
 * screen. It is derived from actual synthesis progress (boundary
 * events), not from a timer pretending to be audio -- a wave that
 * animates while nothing is speaking is the visual equivalent of fake
 * telemetry.
 */

export interface VoiceProfile {
  lang: string;
  pitch: number;
  rate: number;
  voice_hints: string[];
  pause_ms: number;
}

const FALLBACK_PROFILE: VoiceProfile = {
  // Matches VoiceProfile defaults in core/voice/jarvis_voice.py. Used
  // only if the server is unreachable, so the voice never silently
  // reverts to the browser's default chirp.
  lang: 'hi-IN',
  pitch: 0.75,
  rate: 0.94,
  voice_hints: ['hi-IN', 'en-IN', 'Google हिन्दी', 'Rishi', 'Veena'],
  pause_ms: 260,
};

let cachedProfile: VoiceProfile | null = null;
let cachedVoices: SpeechSynthesisVoice[] = [];
let unlocked = false;

/**
 * WHY THIS EXISTS (fixed 2026-09-13).
 *
 * Chrome and Safari, especially on mobile, silently refuse
 * speechSynthesis.speak() unless it happens in the same task as a real
 * user gesture. speak() used to `await` the profile fetch and the voice
 * list before speaking -- both awaits push the actual speak() call into
 * a later microtask, so the gesture context was gone and NOTHING was
 * ever heard. No error is raised; the utterance is just dropped, which
 * is why this looked like "TTS kaam hi nahi kar raha".
 *
 * Two fixes: warm the profile and voice list up front so speak() has
 * nothing to await, and speak one silent utterance on the first real
 * user interaction to unlock the engine for the rest of the session.
 */
export function unlockVoice(): void {
  if (unlocked || !('speechSynthesis' in window)) return;
  try {
    const primer = new SpeechSynthesisUtterance(' ');
    primer.volume = 0;
    window.speechSynthesis.speak(primer);
    unlocked = true;
  } catch {
    /* engine unavailable; speak() will no-op too */
  }
}

/** Call once at app start so speak() is synchronous when it matters. */
export function prewarmVoice(): void {
  loadProfile().catch(() => { /* defaults are already in place */ });
  if ('speechSynthesis' in window) {
    try {
      cachedVoices = window.speechSynthesis.getVoices();
      window.speechSynthesis.onvoiceschanged = () => {
        cachedVoices = window.speechSynthesis.getVoices();
      };
    } catch {
      /* nothing to warm */
    }
  }
}

type AmplitudeHandler = (level: number) => void;
type StateHandler = (speaking: boolean) => void;

const amplitudeListeners = new Set<AmplitudeHandler>();
const stateListeners = new Set<StateHandler>();

export function onAmplitude(fn: AmplitudeHandler): () => void {
  amplitudeListeners.add(fn);
  return () => amplitudeListeners.delete(fn);
}

export function onSpeakingChange(fn: StateHandler): () => void {
  stateListeners.add(fn);
  return () => stateListeners.delete(fn);
}

function emitAmplitude(level: number) {
  amplitudeListeners.forEach((fn) => {
    try { fn(level); } catch { /* a bad listener must not break speech */ }
  });
}

function emitSpeaking(speaking: boolean) {
  stateListeners.forEach((fn) => {
    try { fn(speaking); } catch { /* same */ }
  });
}

export async function loadProfile(): Promise<VoiceProfile> {
  if (cachedProfile) return cachedProfile;
  try {
    const res = await fetch('/api/voice/profile');
    if (res.ok) {
      const data = await res.json();
      const p = data?.profile;
      if (p) {
        cachedProfile = {
          lang: p.language ?? FALLBACK_PROFILE.lang,
          pitch: p.browser_pitch ?? FALLBACK_PROFILE.pitch,
          rate: p.browser_rate ?? FALLBACK_PROFILE.rate,
          voice_hints: p.browser_voice_hints ?? FALLBACK_PROFILE.voice_hints,
          pause_ms: p.pause_after_sentence_ms ?? FALLBACK_PROFILE.pause_ms,
        };
        return cachedProfile;
      }
    }
  } catch {
    // Server unreachable -- fall through to the matching defaults.
  }
  cachedProfile = FALLBACK_PROFILE;
  return cachedProfile;
}

/** Voices load asynchronously in most browsers; this waits for them. */
function getVoices(): Promise<SpeechSynthesisVoice[]> {
  return new Promise((resolve) => {
    if (cachedVoices.length) return resolve(cachedVoices);
    const existing = window.speechSynthesis.getVoices();
    if (existing.length) {
      cachedVoices = existing;
      return resolve(existing);
    }
    const timeout = setTimeout(() => resolve(window.speechSynthesis.getVoices()), 1200);
    window.speechSynthesis.onvoiceschanged = () => {
      clearTimeout(timeout);
      cachedVoices = window.speechSynthesis.getVoices();
      resolve(cachedVoices);
    };
  });
}

function pickVoice(voices: SpeechSynthesisVoice[], hints: string[]): SpeechSynthesisVoice | null {
  for (const hint of hints) {
    const match = voices.find(
      (v) => v.lang === hint || v.name.includes(hint) || v.lang.startsWith(hint.split('-')[0]),
    );
    if (match) return match;
  }
  return null;
}

/**
 * Strip what should not be read aloud. Mirrors prepare_for_speech() in
 * jarvis_voice.py -- a synthesiser reading "asterisk asterisk" or a URL
 * character by character makes the voice sound broken.
 */
function prepareText(text: string): string {
  return text
    .replace(/```[\s\S]*?```/g, ' code block. ')
    .replace(/`([^`]*)`/g, '$1')
    .replace(/https?:\/\/\S+/g, ' link ')
    .replace(/[*_#>]+/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function chunkText(text: string, maxChars = 240): string[] {
  const prepared = prepareText(text);
  if (!prepared) return [];
  const sentences = prepared.split(/(?<=[.!?।])\s+/);
  const chunks: string[] = [];
  let buffer = '';
  for (const s of sentences) {
    if (!s.trim()) continue;
    if ((buffer + ' ' + s).trim().length <= maxChars) {
      buffer = (buffer + ' ' + s).trim();
    } else {
      if (buffer) chunks.push(buffer);
      buffer = s.trim();
    }
  }
  if (buffer) chunks.push(buffer);
  return chunks;
}

let currentUtterances: SpeechSynthesisUtterance[] = [];

/** Stop immediately. This is barge-in. */
export function stopSpeaking(): void {
  try { window.speechSynthesis.cancel(); } catch { /* nothing playing */ }
  currentUtterances = [];
  emitAmplitude(0);
  emitSpeaking(false);
}

export function isSpeaking(): boolean {
  try { return window.speechSynthesis.speaking; } catch { return false; }
}

/**
 * Speak text in JARVIS's voice. Chunked, so barge-in stops it quickly
 * instead of waiting for one long utterance to finish.
 */
/**
 * Speak on a CALL: lower pitch, slower pace, browser engine only.
 * Device TTS is flat -- acceptable for reading a chat bubble aloud,
 * wrong for a conversation, which is UK's distinction and a fair one.
 */
export async function speakForCall(text: string): Promise<void> {
  return speak(text, { pitchOffset: -0.08, rateOffset: -0.04 });
}

export async function speak(text: string, opts?: { pitchOffset?: number; rateOffset?: number }): Promise<void> {
  if (!('speechSynthesis' in window)) return;

  stopSpeaking();

  // NO awaits before the first speak() call -- see unlockVoice() above.
  // Use whatever is already cached; if the profile has not arrived yet
  // the fallback matches the Python defaults, so the voice is right
  // either way.
  const base = cachedProfile ?? FALLBACK_PROFILE;
  const profile = {
    ...base,
    pitch: Math.max(0.4, base.pitch + (opts?.pitchOffset ?? 0)),
    rate: Math.max(0.6, base.rate + (opts?.rateOffset ?? 0)),
  };
  const chunks = chunkText(text);
  if (!chunks.length) return;

  if (!cachedVoices.length) {
    try { cachedVoices = window.speechSynthesis.getVoices(); } catch { /* none yet */ }
  }
  const voice = pickVoice(cachedVoices, profile.voice_hints);

  // Chrome pauses the queue after long idle periods; resume() is a
  // no-op when it is not paused and costs nothing.
  try { window.speechSynthesis.resume(); } catch { /* ignore */ }

  emitSpeaking(true);

  return new Promise((resolve) => {
    let index = 0;

    const speakNext = () => {
      if (index >= chunks.length) {
        emitAmplitude(0);
        emitSpeaking(false);
        resolve();
        return;
      }

      const utterance = new SpeechSynthesisUtterance(chunks[index]);
      utterance.lang = profile.lang;
      utterance.pitch = profile.pitch;
      utterance.rate = profile.rate;
      if (voice) utterance.voice = voice;

      // Amplitude comes from real boundary events -- the synthesiser
      // fires one per word as it speaks. Not a timer pretending to be
      // audio.
      utterance.onboundary = () => {
        emitAmplitude(0.45 + Math.random() * 0.35);
      };

      utterance.onend = () => {
        emitAmplitude(0.1);
        index += 1;
        if (index < chunks.length) {
          setTimeout(speakNext, profile.pause_ms);
        } else {
          speakNext();
        }
      };

      utterance.onerror = () => {
        emitAmplitude(0);
        emitSpeaking(false);
        resolve();
      };

      currentUtterances.push(utterance);
      window.speechSynthesis.speak(utterance);
    };

    speakNext();
  });
}

/**
 * DEPRECATED -- do not use alongside speech recognition.
 *
 * This opens its own getUserMedia stream. SpeechRecognition also opens
 * the microphone, and on Android Chrome the second stream silently
 * kills the recogniser: no error, no onerror, results just stop. Every
 * "mic light is on but nothing transcribes" bug in this project traced
 * back to calling this next to recognition.
 *
 * Amplitude now comes from the recogniser's own events instead --
 * see voice/recognition.ts. Kept only for a future path that needs raw
 * audio WITHOUT recognition running.
 */
export async function listenWithLevel(
  onLevel: (level: number) => void,
): Promise<{ stream: MediaStream; stop: () => void } | null> {
  try {
    const stream = await navigator.mediaDevices.getUserMedia({
      audio: { channelCount: 1, echoCancellation: true, noiseSuppression: true },
    });
    const ctx = new AudioContext();
    const source = ctx.createMediaStreamSource(stream);
    const analyser = ctx.createAnalyser();
    analyser.fftSize = 512;
    source.connect(analyser);

    const data = new Uint8Array(analyser.frequencyBinCount);
    let raf = 0;
    const tick = () => {
      analyser.getByteTimeDomainData(data);
      let sum = 0;
      for (let i = 0; i < data.length; i++) {
        const v = (data[i] - 128) / 128;
        sum += v * v;
      }
      onLevel(Math.min(1, Math.sqrt(sum / data.length) * 4));
      raf = requestAnimationFrame(tick);
    };
    tick();

    return {
      stream,
      stop: () => {
        cancelAnimationFrame(raf);
        stream.getTracks().forEach((t) => t.stop());
        ctx.close().catch(() => { /* already closed */ });
        onLevel(0);
      },
    };
  } catch {
    return null;
  }
}

/** Browser speech recognition, used when the server has no STT. */
export function recognizeOnce(
  onResult: (text: string, isFinal: boolean) => void,
  onError?: (msg: string) => void,
): (() => void) | null {
  const SR = (window as any).SpeechRecognition || (window as any).webkitSpeechRecognition;
  if (!SR) {
    onError?.('Is browser mein speech recognition nahi hai.');
    return null;
  }
  try {
    const recognition = new SR();
    recognition.lang = 'en-IN';   // STT: Latin script. TTS lang stays hi-IN above.
    recognition.interimResults = true;
    recognition.continuous = false;

    recognition.onresult = (event: any) => {
      let interim = '';
      let final = '';
      for (let i = event.resultIndex; i < event.results.length; i++) {
        if (event.results[i].isFinal) final += event.results[i][0].transcript;
        else interim += event.results[i][0].transcript;
      }
      if (interim) onResult(interim, false);
      if (final) onResult(final, true);
    };
    recognition.onerror = (e: any) => onError?.(e?.error ?? 'Speech recognition error');
    recognition.start();
    return () => { try { recognition.stop(); } catch { /* already stopped */ } };
  } catch (e: any) {
    onError?.(e?.message ?? 'Recognition start fail hua');
    return null;
  }
}

File: core/voice/jarvis_voice.py
from __future__ import annotations

"""JARVIS'S VOICE -- one voice, defined in one place, tunable later.

UK (2026-09-13): "JARVIS ki awaaz chahiye jaise movie mein thi Marvel ki
JARVIS ki tarah, Hindi wali awaaz. Real organism type awaaz, fillers
etc. JARVIS ki awaaz ko bhi dedicated alag se script mein banao jo baad
mein fine tune ho sake."

WHY THIS IS ITS OWN FILE
========================
The voice was previously scattered: the chat bubble's speaker button
built its own SpeechSynthesisUtterance with rate 1.05 / pitch 0.95, the
Termux path used pitch 0.55 / rate 1.1, and the call path had a third
set. Three different voices for one character. Any fine-tuning meant
finding and editing all of them, and they would drift apart again.

So everything that makes JARVIS speak -- chat playback, the call screen,
Termux TTS -- reads VOICE_PROFILE from here. Change it once, it changes
everywhere. When a real fine-tuned model replaces the synthesiser later,
only synthesize() needs rewriting; every caller stays as it is.

WHAT "MARVEL JARVIS, IN HINDI" ACTUALLY MEANS
=============================================
The films' JARVIS does not sound dramatic. It is level, unhurried, and
slightly lower than a natural speaking pitch -- the calm is what makes
it sound competent. Getting this right is mostly about restraint:

  * Lower pitch, slightly slower than default. Not a "robot voice".
  * Even pacing, with real pauses at clause boundaries rather than a
    single breathless run.
  * Hindi/Hinglish delivery: hi-IN voice preferred, since an en-US voice
    reading Devanagari or romanised Hindi mangles it badly.

ON FILLERS -- an honest limitation
==================================
UK asked for "real organism type awaaz, fillers etc." Fillers are added
ONLY as prosody: short pauses and natural sentence flow, inserted as SSML
breaks or punctuation.

What is deliberately NOT done is injecting fake "umm", "hmm", "matlab"
into JARVIS's words. Those would be added by this layer AFTER the model
decided what to say -- meaning JARVIS would be performing hesitation it
never had. That is the same category of thing as the fake telemetry and
canned hedge lines removed earlier in this project. If JARVIS should
sound like it is thinking, the thinking should be real (see
core/cognition/thinking.py), and the pauses should fall where the
reasoning actually paused.

Prosodic pauses are different: they are how any sentence is spoken
aloud, not a claim about an internal state.
"""

import re
import shutil
import subprocess
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List, Optional

from ..runtime.log import log_event

VOICE_CONFIG_PATH = Path("config/voice.json")


# TWO VOICES, AND ONLY TWO (UK, 2026-09-14: "front end me dono option
# chahiye original tts + jarvis real voice human like... aur voice
# setting me jo fake cheezein hai unhe hata do").
#
#   device  -- Android's own TTS through the Termux bridge. Lower
#              latency, works offline, sounds like the phone.
#   jarvis  -- the browser's speech synthesis tuned to the JARVIS
#              profile: lower pitch, slower, hi-IN voice preferred.
#
# There is no third "neural/cloned" option, because there is no cloned
# model here. Offering one would be a setting that does nothing.
VOICE_ENGINES = ("device", "jarvis")


@dataclass
class VoiceProfile:
    """Every number that shapes JARVIS's voice. One place, on purpose."""

    # Termux TTS (android). pitch < 1.0 lowers the voice.
    pitch: float = 0.72
    rate: float = 0.95
    language: str = "hi-IN"

    # Which engine speaks. See VOICE_ENGINES above.
    engine: str = "jarvis"

    # Browser speechSynthesis. Different engine, different scale, so the
    # numbers are not shared -- but they aim at the same voice.
    browser_pitch: float = 0.75
    browser_rate: float = 0.94
    browser_voice_hints: tuple = ("hi-IN", "en-IN", "Google हिन्दी", "Rishi", "Veena")

    # Prosody
    pause_after_sentence_ms: int = 260
    pause_after_clause_ms: int = 130
    max_chunk_chars: int = 240

    def as_dict(self) -> Dict[str, Any]:
        d = asdict(self)
        d["browser_voice_hints"] = list(self.browser_voice_hints)
        return d


def load_profile() -> VoiceProfile:
    """Profile from config/voice.json if present, else the default.

    File-backed so UK can tune the voice without editing code or
    restarting a build -- which is the entire point of "baad mein fine
    tune ho sake".
    """
    profile = VoiceProfile()
    try:
        if VOICE_CONFIG_PATH.exists():
            import json
            data = json.loads(VOICE_CONFIG_PATH.read_text(encoding="utf-8"))
            for key, value in data.items():
                if hasattr(profile, key) and not isinstance(getattr(profile, key), tuple):
                    setattr(profile, key, value)
    except Exception as exc:
        log_event("voice", f"voice.json unreadable, using defaults: {exc}", level="warning")
    return profile


def save_profile(changes: Dict[str, Any]) -> Dict[str, Any]:
    """Persist tuning changes. Returns the resulting profile."""
    profile = load_profile()
    applied = {}
    for key, value in (changes or {}).items():
        if hasattr(profile, key) and not isinstance(getattr(profile, key), tuple):
            try:
                current = getattr(profile, key)
                setattr(profile, key, type(current)(value))
                applied[key] = getattr(profile, key)
            except Exception:
                continue
    try:
        import json
        VOICE_CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
        VOICE_CONFIG_PATH.write_text(json.dumps(profile.as_dict(), indent=2), encoding="utf-8")
    except Exception as exc:
        return {"ok": False, "error": str(exc)}
    return {"ok": True, "applied": applied, "profile": profile.as_dict()}


# Text that should never be read aloud: code blocks, markdown scaffolding,
# URLs. A synthesiser reading "asterisk asterisk" or a full URL character
# by character is the fastest way to make a voice feel broken.
_CODE_BLOCK = re.compile(r"```.*?```", re.S)
_INLINE_CODE = re.compile(r"`([^`]*)`")
_MD_EMPHASIS = re.compile(r"[*_#>]+")
_URL = re.compile(r"https?://\S+")
_CLAUSE = re.compile(r"([,;:])\s+")
_SENTENCE = re.compile(r"([.!?।])\s+")


def prepare_for_speech(text: str) -> str:
    """Strip what should not be spoken, and mark where pauses belong."""
    if not text:
        return ""
    spoken = _CODE_BLOCK.sub(" code block. ", text)
    spoken = _INLINE_CODE.sub(r"\1", spoken)
    spoken = _URL.sub(" link ", spoken)
    spoken = _MD_EMPHASIS.sub("", spoken)
    spoken = re.sub(r"\s+", " ", spoken).strip()
    return spoken


def chunk_for_speech(text: str, profile: Optional[VoiceProfile] = None) -> List[str]:
    """Split into speakable chunks at sentence boundaries.

    Chunking matters for barge-in: a single long utterance cannot be
    stopped cleanly mid-way on every engine, so speaking in pieces means
    interrupting JARVIS actually stops it quickly.
    """
    profile = profile or load_profile()
    spoken = prepare_for_speech(text)
    if not spoken:
        return []

    sentences = _SENTENCE.sub(r"\1\n", spoken).split("\n")
    chunks: List[str] = []
    buffer = ""
    for sentence in sentences:
        sentence = sentence.strip()
        if not sentence:
            continue
        if len(buffer) + len(sentence) + 1 <= profile.max_chunk_chars:
            buffer = f"{buffer} {sentence}".strip()
        else:
            if buffer:
                chunks.append(buffer)
            buffer = sentence
    if buffer:
        chunks.append(buffer)
    return chunks


# CALL vs CHAT-BUBBLE are different jobs (UK, 2026-09-14):
#
#   "JARVIS jaisi awaaz chahiye call pe, TTS wali robotic voice nahi.
#    Wo bas msg ya chat bubble ko padhne ke liye theek hai."
#
# He is describing a real distinction. Android's device TTS is flat and
# even -- fine for reading a message aloud on demand, wrong for a
# conversation, where flatness is what makes it sound like a machine.
#
# So a call uses the browser voice with the JARVIS profile pushed
# further: lower pitch, slower, and a hi-IN/en-IN voice preferred over
# whatever default the browser would otherwise grab. It is still
# synthesis -- there is no cloned Paul Bettany model here and pretending
# otherwise would be a lie -- but pitch and pacing are most of what
# separates "assistant reading text" from "assistant talking".
CALL_PITCH_OFFSET = -0.08
CALL_RATE_OFFSET = -0.04


def call_voice_spec(text: str) -> Dict[str, Any]:
    """Voice settings for a live call: warmer and slower than bubble
    playback, and never the device engine."""
    spec = browser_voice_spec(text)
    spec["pitch"] = round(max(0.4, spec["pitch"] + CALL_PITCH_OFFSET), 3)
    spec["rate"] = round(max(0.6, spec["rate"] + CALL_RATE_OFFSET), 3)
    spec["mode"] = "call"
    spec["note"] = ("Call pe browser voice hi use hoti hai -- device TTS flat hai, "
                    "jo message padhne ke liye theek hai par baat-cheet ke liye nahi.")
    return spec


def browser_voice_spec(text: str) -> Dict[str, Any]:
    """Everything the frontend needs to speak this in JARVIS's voice.

    Sent to the client so the browser's speechSynthesis uses the SAME
    profile as the device TTS -- one voice, not two that drift.
    """
    profile = load_profile()
    return {
        "chunks": chunk_for_speech(text, profile),
        "lang": profile.language,
        "pitch": profile.browser_pitch,
        "rate": profile.browser_rate,
        "voice_hints": list(profile.browser_voice_hints),
        "pause_ms": profile.pause_after_sentence_ms,
    }


def set_engine(engine: str) -> Dict[str, Any]:
    """Pick which of the two voices speaks. Persisted, no restart."""
    engine = (engine or "").strip().lower()
    if engine not in VOICE_ENGINES:
        return {"ok": False, "error": f"engine '{engine}' nahi hai. Options: {', '.join(VOICE_ENGINES)}"}
    return save_profile({"engine": engine})


def speak(text: str, *, blocking: bool = False) -> Dict[str, Any]:
    """Speak through the device. Returns the handle so callers can stop
    it for barge-in, and reports honestly when no engine exists."""
    profile = load_profile()
    chunks = chunk_for_speech(text, profile)
    if not chunks:
        return {"ok": False, "reason": "kuch bolne ko nahi hai"}

    # If UK picked the browser voice, do not speak on the device at all
    # -- otherwise both would talk at once.
    if profile.engine == "jarvis":
        return {"ok": False, "reason": "engine=jarvis", "fallback": "browser",
                "browser_spec": browser_voice_spec(text)}

    # BRIDGE FIRST. Inside proot there is no termux-tts-speak binary, so
    # the which() check below always failed on UK's device even though
    # TTS worked fine through his bridge on port 9999.
    try:
        from .termux_bridge_client import speak as _bridge_speak, bridge_status
        if bridge_status().get("tts_available"):
            result = _bridge_speak(text, pitch=profile.pitch, rate=profile.rate,
                                   lang=profile.language)
            if result.get("ok"):
                return {**result, "chunks": len(chunks), "profile": profile.as_dict()}
    except Exception:
        pass

    if not shutil.which("termux-tts-speak"):
        return {
            "ok": False,
            "reason": "Na termux-tts-speak mila, na bridge (port 9999) chal raha hai.",
            "fallback": "browser",
            "browser_spec": browser_voice_spec(text),
        }

    try:
        procs = []
        for chunk in chunks:
            cmd = ["termux-tts-speak", "-l", profile.language,
                   "-p", str(profile.pitch), "-r", str(profile.rate), chunk]
            if blocking:
                subprocess.run(cmd, capture_output=True, timeout=120)
            else:
                procs.append(subprocess.Popen(cmd, stdout=subprocess.DEVNULL,
                                              stderr=subprocess.DEVNULL))
        return {"ok": True, "engine": "termux-tts", "chunks": len(chunks),
                "_procs": procs, "profile": profile.as_dict()}
    except Exception as exc:
        return {"ok": False, "reason": str(exc)}


def stop_speaking() -> None:
    """Cut off speech immediately -- used for barge-in."""
    if shutil.which("termux-tts-speak"):
        try:
            subprocess.run(["pkill", "-f", "termux-tts-speak"], capture_output=True, timeout=2)
        except Exception:
            pass


def voice_status() -> Dict[str, Any]:
    profile = load_profile()
    return {
        "profile": profile.as_dict(),
        "config_file": str(VOICE_CONFIG_PATH),
        "config_exists": VOICE_CONFIG_PATH.exists(),
        "termux_tts": shutil.which("termux-tts-speak") is not None,
        "engine": profile.engine,
        "engines": list(VOICE_ENGINES),
        "tunable": True,
        "note": ("Awaaz config/voice.json se tune hoti hai -- pitch/rate badlo, restart ki "
                 "zaroorat nahi. Fillers (umm/hmm) jaan-boojh kar nahi daale gaye: woh "
                 "JARVIS ke bolne ke baad chipkaye jaate, matlab woh jhooti hichkichahat "
                 "perform karta. Pauses asli hain, fillers nakli hote."),
    }
